File: src/claudestruct/tracing.py
"""OpenTelemetry tracing — opt-in, zero-cost when disabled.

Activates only when `OTEL_EXPORTER_OTLP_ENDPOINT` is set in the
environment (matching the OTel convention). Without it, every helper
in this module is a no-op so unrelated runs pay no overhead and the
heavy `opentelemetry` SDK never loads.

Wired into:
  - `runner.run_task_and_log` — root span "claudestruct.task" with
    `task`, `model`, `prompt_version` attributes; child span for the
    actual `run_task` call carrying input/output token counts.

The OTel SDK is a soft dep: declared as `claudestruct[otel]` extra in
pyproject.toml. When the user enables tracing without installing the
extra, we emit a single warning to stderr and fall back to no-op
spans so the run continues unaffected.

Why OTLP/HTTP and not OTLP/gRPC: HTTP works through restrictive proxies
that block grpc, has no protobuf-compiler runtime cost, and the perf
gap is negligible for the low-volume traces this CLI emits.
"""
from __future__ import annotations

import contextlib
import os
import sys
from collections.abc import Iterator
from typing import Any


class _NoopSpan:
    """Stub span returned when OTel is disabled. Exists so callers can
    always do `with tracing.span(...) as s: s.set_attribute(...)`
    without a None check."""

    def set_attribute(self, _key: str, _value: Any) -> None:
        pass

    def record_exception(self, _exc: BaseException) -> None:
        pass


_NOOP_SPAN = _NoopSpan()


# Module-level flag set on first init_tracing(). Re-init is idempotent.
_initialized: bool = False
_enabled: bool = False
_tracer: Any = None


def is_enabled() -> bool:
    return _enabled


def init_tracing(service_name: str = "claudestruct") -> bool:
    """Initialise tracing if `OTEL_EXPORTER_OTLP_ENDPOINT` is set.

    Returns True if tracing is now active, False otherwise. Idempotent:
    a second call returns the prior result without re-initializing.
    """
    global _initialized, _enabled, _tracer
    if _initialized:
        return _enabled
    _initialized = True

    endpoint = os.environ.get("OTEL_EXPORTER_OTLP_ENDPOINT", "").strip()
    if not endpoint:
        _enabled = False
        return False

    try:
        from opentelemetry import trace
        from opentelemetry.exporter.otlp.proto.http.trace_exporter import (
            OTLPSpanExporter,
        )
        from opentelemetry.sdk.resources import Resource
        from opentelemetry.sdk.trace import TracerProvider
        from opentelemetry.sdk.trace.export import BatchSpanProcessor
    except ImportError:
        # Soft dep: don't crash the run, just warn once. Users who
        # want tracing should `pip install claudestruct[otel]`.
        sys.stderr.write(
            "[tracing] OTEL_EXPORTER_OTLP_ENDPOINT is set but the "
            "opentelemetry SDK is not installed. Run "
            "`pip install opentelemetry-api opentelemetry-sdk "
            "opentelemetry-exporter-otlp-proto-http` "
            "(or `pip install claudestruct[otel]`) to enable tracing.\n"
        )
        _enabled = False
        return False

    resource = Resource.create({"service.name": service_name})
    provider = TracerProvider(resource=resource)
    # The OTLP HTTP exporter reads OTEL_EXPORTER_OTLP_ENDPOINT itself,
    # plus standard OTEL_EXPORTER_OTLP_HEADERS for auth tokens.
    exporter = OTLPSpanExporter()
    provider.add_span_processor(BatchSpanProcessor(exporter))
    trace.set_tracer_provider(provider)
    _tracer = trace.get_tracer(service_name)
    _enabled = True
    return True


@contextlib.contextmanager
def span(name: str, attributes: dict[str, Any] | None = None) -> Iterator[Any]:
    """Open a span. Returns a no-op object when tracing is disabled,
    so callers can use the same `with` block unconditionally."""
    if not _enabled or _tracer is None:
        yield _NOOP_SPAN
        return
    with _tracer.start_as_current_span(name) as s:
        if attributes:
            for k, v in attributes.items():
                # OTel only accepts primitives + lists of primitives.
                # Anything else gets stringified — never raise on
                # unexpected types, that's worse than a string label.
                if isinstance(v, (str, int, float, bool)):
                    s.set_attribute(k, v)
                elif v is None:
                    pass
                else:
                    s.set_attribute(k, str(v))
        try:
            yield s
        except BaseException as exc:
            s.record_exception(exc)
            raise


def shutdown() -> None:
    """Flush pending spans. Call before process exit so the BatchSpanProcessor
    doesn't drop in-flight traces. No-op when tracing is disabled."""
    if not _enabled:
        return
    try:
        from opentelemetry import trace
        provider = trace.get_tracer_provider()
        flush = getattr(provider, "shutdown", None)
        if callable(flush):
            flush()
    except Exception:
        # Shutdown is best-effort: a flush failure must not mask the
        # actual run's exit code or stop the process from exiting.
        pass

File: claw-squad/src/tracing.ts
/**
 * OpenTelemetry tracing for claw-squad — opt-in, zero-cost when disabled.
 *
 * Mirrors the Python side (`src/claudestruct/tracing.py`) so a single
 * OTel collector can ingest traces from both tools using the same
 * span-name conventions:
 *
 *   - `clawSquad.run`            (root span, one per `runOrchestrator`)
 *     - `clawSquad.planner.call`
 *     - `clawSquad.coder.call`
 *     - `clawSquad.reviewer.call`
 *     - `clawSquad.subagent.call` (with `subagent.name` attribute)
 *
 * Activates only when `OTEL_EXPORTER_OTLP_ENDPOINT` is set. Without it
 * every helper is a no-op and the OTel SDK never loads. The SDK is
 * declared as an `optionalDependencies` block in package.json so a
 * normal `pnpm install` skips it; users who want tracing run
 * `pnpm install @opentelemetry/api @opentelemetry/sdk-node \
 *  @opentelemetry/exporter-trace-otlp-http`.
 *
 * Why HTTP exporter (not gRPC): same reasoning as the Python module —
 * proxies / CDNs frequently block gRPC; HTTP is universally usable
 * and the perf gap is negligible for the CLI's trace volume.
 */

interface Span {
  setAttribute(key: string, value: string | number | boolean): void;
  end(): void;
  recordException(err: Error): void;
}

const NOOP_SPAN: Span = {
  setAttribute: () => {},
  end: () => {},
  recordException: () => {},
};

let initialized = false;
let enabled = false;
let tracer: { startSpan(name: string): Span } | null = null;
let shutdownFn: (() => Promise<void>) | null = null;

export function isEnabled(): boolean {
  return enabled;
}

/**
 * Initialise tracing if the OTLP endpoint is set in env. Idempotent;
 * a second call returns the same enabled flag without re-init.
 *
 * Returns a Promise<boolean> because the OTel SDK's dynamic `import`
 * is async. Callers in synchronous orchestration contexts can await
 * this once at startup; the helpers below stay synchronous.
 */
export async function initTracing(serviceName = "claw-squad"): Promise<boolean> {
  if (initialized) return enabled;
  initialized = true;

  const endpoint = (process.env.OTEL_EXPORTER_OTLP_ENDPOINT ?? "").trim();
  if (!endpoint) {
    enabled = false;
    return false;
  }

  try {
    // Dynamic import keeps the SDK out of the cold-start cost when
    // tracing is disabled, AND lets `pnpm install` skip the optional
    // dep without breaking compilation. Marked `// @ts-ignore` because
    // the package isn't part of the always-installed type surface.
    // @ts-ignore — optional peer
    const api = await import("@opentelemetry/api");
    // @ts-ignore — optional peer
    const sdkNode = await import("@opentelemetry/sdk-node");
    // @ts-ignore — optional peer
    const otlpHttp = await import("@opentelemetry/exporter-trace-otlp-http");

    const sdk = new sdkNode.NodeSDK({
      serviceName,
      traceExporter: new otlpHttp.OTLPTraceExporter(),
    });
    sdk.start();
    shutdownFn = () => sdk.shutdown();

    const apiTracer = api.trace.getTracer(serviceName);
    tracer = {
      startSpan(name: string): Span {
        const s = apiTracer.startSpan(name);
        return {
          setAttribute(k, v) {
            s.setAttribute(k, v);
          },
          end() {
            s.end();
          },
          recordException(err) {
            s.recordException(err);
          },
        };
      },
    };
    enabled = true;
    return true;
  } catch (err) {
    // Soft dep: don't crash the run, warn once. Users who want tracing
    // should install the OTel packages explicitly.
    process.stderr.write(
      `[tracing] OTEL_EXPORTER_OTLP_ENDPOINT is set but the OpenTelemetry ` +
        `SDK is not installed. Run \`pnpm add @opentelemetry/api ` +
        `@opentelemetry/sdk-node @opentelemetry/exporter-trace-otlp-http\` ` +
        `to enable tracing.\n` +
        (err instanceof Error ? `[tracing] reason: ${err.message}\n` : ""),
    );
    enabled = false;
    return false;
  }
}

/**
 * Open a span, run the function, end the span. Returns whatever `fn`
 * returns. Exceptions are recorded on the span before re-throwing so
 * the trace shows the failure mode.
 */
export async function withSpan<T>(
  name: string,
  attrs: Record<string, string | number | boolean | undefined>,
  fn: (span: Span) => Promise<T>,
): Promise<T> {
  if (!enabled || tracer === null) {
    return fn(NOOP_SPAN);
  }
  const span = tracer.startSpan(name);
  for (const [k, v] of Object.entries(attrs)) {
    if (v === undefined) continue;
    span.setAttribute(k, v);
  }
  try {
    return await fn(span);
  } catch (err) {
    if (err instanceof Error) span.recordException(err);
    throw err;
  } finally {
    span.end();
  }
}

/** Flush + shut down the trace pipeline. Call before process exit. */
export async function shutdownTracing(): Promise<void> {
  if (!enabled || shutdownFn === null) return;
  try {
    await shutdownFn();
  } catch {
    // Best-effort: shutdown failure must not mask the real exit code.
  }
}

// ---- Test-only helpers ---------------------------------------------
//
// `_resetForTests` lets vitest reset module-level state between cases
// so the disabled-by-default contract is testable without process
// recycling. Not exported through index; intended for tests only.

export function _resetForTests(): void {
  initialized = false;
  enabled = false;
  tracer = null;
  shutdownFn = null;
}

export function _injectForTests(t: { startSpan(name: string): Span } | null): void {
  tracer = t;
  enabled = t !== null;
  initialized = true;
}
